core/drivers/crypto/hisilicon/hpre_rsa.c

```diff
@@ -0,0 +1,1076 @@
+		return ret;
+	}
+
+	ret = hpre_bin_from_crypto_bin(dp, dp, p_bytes, dp_len);
+	if (ret) {
+		EMSG("Fail to transfer rsa crt dp from crypto_bin to hpre_bin");
+		return ret;
+	}
+
+	ret = hpre_bin_from_crypto_bin(q, q, p_bytes, q_len);
+	if (ret) {
+		EMSG("Fail to transfer rsa crt q from crypto_bin to hpre_bin");
+		return ret;
+	}
+
+	ret = hpre_bin_from_crypto_bin(p, p, p_bytes, p_len);
+	if (ret) {
+		EMSG("Fail to transfer rsa crt p from crypto_bin to hpre_bin");
+		return ret;
+	}
+
+	ret = hpre_bin_from_crypto_bin(qp, qp, p_bytes, qp_len);
+	if (ret) {
+		EMSG("Fail to transfer rsa crt qinv from crypto_bin to hpre_bin");
+		return ret;
+	}
+
+	ret = hpre_bin_from_crypto_bin(msg->in, rsa_data->cipher.data,
+				       msg->key_bytes, rsa_data->cipher.length);
+	if (ret)
+		EMSG("Fail to transfer rsa ciphertext from crypto_bin to hpre_bin");
+
+	return ret;
+}
+
+static enum hisi_drv_status
+hpre_rsa_ncrt_decrypt_bn2bin(struct hpre_rsa_msg *msg,
+			     struct drvcrypt_rsa_ed *rsa_data)
+{
+	struct rsa_keypair *key = rsa_data->key.key;
+	uint32_t d_len = 0;
+	uint32_t n_len = 0;
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+	uint8_t *n = NULL;
+
+	n = msg->prikey + msg->key_bytes;
+
+	crypto_bignum_bn2bin(key->d, msg->prikey);
+	crypto_bignum_bn2bin(key->n, n);
+	d_len = crypto_bignum_num_bytes(key->d);
+	n_len = crypto_bignum_num_bytes(key->n);
+
+	ret = hpre_bin_from_crypto_bin(msg->prikey, msg->prikey, msg->key_bytes,
+				       d_len);
+	if (ret) {
+		EMSG("Fail to transfer rsa ncrt d from crypto_bin to hpre_bin");
+		return ret;
+	}
+
+	ret = hpre_bin_from_crypto_bin(n, n, msg->key_bytes, n_len);
+	if (ret) {
+		EMSG("Fail to transfer rsa ncrt n from crypto_bin to hpre_bin");
+		return ret;
+	}
+
+	ret = hpre_bin_from_crypto_bin(msg->in, rsa_data->cipher.data,
+				       msg->key_bytes, rsa_data->cipher.length);
+	if (ret)
+		EMSG("Fail to transfer rsa ciphertext from crypto_bin to hpre_bin");
+
+	return ret;
+}
+
+static int32_t hpre_rsa_is_crt_mod(struct rsa_keypair *key)
+{
+	int32_t is_crt = 0;
+
+	if (key->p && crypto_bignum_num_bits(key->p) &&
+	    key->q && crypto_bignum_num_bits(key->q) &&
+	    key->dp && crypto_bignum_num_bits(key->dp) &&
+	    key->dq && crypto_bignum_num_bits(key->dq) &&
+	    key->qp && crypto_bignum_num_bits(key->qp))
+		is_crt = 1;
+
+	return is_crt;
+}
+
+static TEE_Result hpre_rsa_decrypt_init(struct hpre_rsa_msg *msg,
+					struct drvcrypt_rsa_ed *rsa_data)
+{
+	struct rsa_keypair *key = rsa_data->key.key;
+	size_t n_bytes = rsa_data->key.n_size;
+	int32_t is_crt = 0;
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+
+	msg->is_private = rsa_data->key.isprivate;
+	msg->key_bytes = hpre_rsa_get_hw_kbytes(BYTES_TO_BITS(n_bytes));
+	if (!msg->key_bytes)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	is_crt = hpre_rsa_is_crt_mod(key);
+	if (is_crt) {
+		msg->alg_type = HPRE_ALG_NC_CRT;
+		ret = hpre_rsa_crt_decrypt_alloc(msg);
+		if (ret)
+			return TEE_ERROR_BAD_STATE;
```

Comment:
Why `TEE_ERROR_BAD_STATE` and not `TEE_ERROR_OUT_OF_MEMORY`?

Reply:
solved

Reply:
You missed this comment.